File: typescript/05-herramientas/23-mcp.ts
// Cliente MCP mínimo.
//
// Implementa el protocolo MCP (Model Context Protocol) via stdio
// para conectar a un servidor MCP, listar sus herramientas y ejecutar tool calls.
//
// El protocolo es JSON-RPC 2.0. El flujo completo:
//   1. initialize (negociar capabilities)
//   2. tools/list (descubrir herramientas disponibles)
//   3. tools/call (ejecutar una herramienta)
//
// Este ejemplo usa un servidor MCP in-process simulado via stdio
// para demostrar el protocolo sin dependencias externas.
// En producción, se usa @modelcontextprotocol/sdk con un servidor real.

// Cómo ejecutar: make ts SCRIPT=typescript/05-herramientas/23-mcp.ts

import Anthropic from "@anthropic-ai/sdk";
import { spawn, ChildProcess } from "child_process";
import { createInterface } from "readline";

const MODEL = process.env["MODEL"] ?? "claude-sonnet-4-6";

// --- Tipos JSON-RPC 2.0 ---

interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number;
  method: string;
  params?: Record<string, unknown>;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: number;
  result?: unknown;
  error?: { code: number; message: string };
}

interface McpTool {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
}

// --- Cliente MCP mínimo via stdio ---

class McpStdioClient {
  private proc: ChildProcess;
  private rl: ReturnType<typeof createInterface>;
  private pending: Map<number, (resp: JsonRpcResponse) => void> = new Map();
  private nextId = 1;

  constructor(command: string, args: string[]) {
    this.proc = spawn(command, args, {
      stdio: ["pipe", "pipe", "inherit"],
    });

    this.rl = createInterface({ input: this.proc.stdout! });
    this.rl.on("line", (line) => {
      if (!line.trim()) return;
      try {
        const msg = JSON.parse(line) as JsonRpcResponse;
        const resolve = this.pending.get(msg.id);
        if (resolve) {
          this.pending.delete(msg.id);
          resolve(msg);
        }
      } catch {
        // ignorar líneas que no son JSON (e.g. logs del servidor)
      }
    });
  }

  private send(method: string, params?: Record<string, unknown>): Promise<unknown> {
    const id = this.nextId++;
    const req: JsonRpcRequest = { jsonrpc: "2.0", id, method, params };
    return new Promise((resolve, reject) => {
      this.pending.set(id, (resp) => {
        if (resp.error) reject(new Error(resp.error.message));
        else resolve(resp.result);
      });
      this.proc.stdin!.write(JSON.stringify(req) + "\n");
    });
  }

  async initialize(): Promise<void> {
    await this.send("initialize", {
      protocolVersion: "2025-03-26",
      capabilities: { tools: {} },
      clientInfo: { name: "agente-libro", version: "1.0" },
    });
    // Notificación (no espera respuesta)
    this.proc.stdin!.write(
      JSON.stringify({ jsonrpc: "2.0", method: "notifications/initialized" }) + "\n"
    );
  }

  async listTools(): Promise<McpTool[]> {
    const result = await this.send("tools/list", {}) as { tools: McpTool[] };
    return result.tools ?? [];
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<unknown> {
    return this.send("tools/call", { name, arguments: args });
  }

  close(): void {
    this.proc.kill();
    this.rl.close();
  }
}

// --- Servidor MCP simulado (in-process via stdio) ---
// En un script separado para poder lanzarlo como subprocess

const SERVER_SCRIPT = `
const readline = require("readline");

const rl = readline.createInterface({ input: process.stdin });
let initialized = false;

const TOOLS = [
  {
    name: "echo",
    description: "Devuelve el texto recibido tal cual.",
    inputSchema: {
      type: "object",
      properties: { text: { type: "string" } },
      required: ["text"]
    }
  },
  {
    name: "add",
    description: "Suma dos números.",
    inputSchema: {
      type: "object",
      properties: {
        a: { type: "number" },
        b: { type: "number" }
      },
      required: ["a", "b"]
    }
  }
];

rl.on("line", (line) => {
  if (!line.trim()) return;
  let req;
  try { req = JSON.parse(line); } catch { return; }

  if (req.method === "initialize") {
    initialized = true;
    const resp = {
      jsonrpc: "2.0", id: req.id,
      result: {
        protocolVersion: "2025-03-26",
        capabilities: { tools: {} },
        serverInfo: { name: "echo-server", version: "1.0" }
      }
    };
    process.stdout.write(JSON.stringify(resp) + "\\n");
    return;
  }

  if (req.method === "notifications/initialized") return;

  if (req.method === "tools/list") {
    process.stdout.write(JSON.stringify({
      jsonrpc: "2.0", id: req.id, result: { tools: TOOLS }
    }) + "\\n");
    return;
  }

  if (req.method === "tools/call") {
    const { name, arguments: args } = req.params;
    let content;
    if (name === "echo") {
      content = [{ type: "text", text: args.text }];
    } else if (name === "add") {
      content = [{ type: "text", text: String(args.a + args.b) }];
    } else {
      process.stdout.write(JSON.stringify({
        jsonrpc: "2.0", id: req.id,
        error: { code: -32601, message: "Tool not found: " + name }
      }) + "\\n");
      return;
    }
    process.stdout.write(JSON.stringify({
      jsonrpc: "2.0", id: req.id,
      result: { content, isError: false }
    }) + "\\n");
    return;
  }

  process.stdout.write(JSON.stringify({
    jsonrpc: "2.0", id: req.id,
    error: { code: -32601, message: "Method not found: " + req.method }
  }) + "\\n");
});
`;

// --- Convertir herramientas MCP al formato de Anthropic ---

function mcpToolToAnthropicTool(tool: McpTool): Anthropic.Tool {
  return {
    name: tool.name,
    description: tool.description,
    input_schema: tool.inputSchema as Anthropic.Tool["input_schema"],
  };
}

// --- Demo completa ---

async function main() {
  console.log("=== Cliente MCP mínimo ===\n");

  // Lanzar el servidor MCP simulado como subprocess
  const fs = await import("fs");
  const os = await import("os");
  const path = await import("path");

  const tmpFile = path.join(os.tmpdir(), "mcp-echo-server.js");
  fs.writeFileSync(tmpFile, SERVER_SCRIPT);

  const mcp = new McpStdioClient("node", [tmpFile]);

  try {
    // 1. Inicializar conexión
    console.log("1. Inicializando conexión MCP...");
    await mcp.initialize();
    console.log("   OK — handshake completado\n");

    // 2. Listar herramientas disponibles
    console.log("2. Descubriendo herramientas (tools/list)...");
    const tools = await mcp.listTools();
    console.log(`   ${tools.length} herramientas disponibles:`);
    for (const tool of tools) {
      console.log(`   - ${tool.name}: ${tool.description}`);
    }
    console.log();

    // 3. Ejecutar una tool call directa via MCP
    console.log("3. Llamada directa: add(17, 25)");
    const addResult = await mcp.callTool("add", { a: 17, b: 25 }) as {
      content: Array<{ type: string; text: string }>;
    };
    console.log(`   Resultado: ${addResult.content[0]?.text}\n`);

    // 4. Usar las herramientas MCP con el modelo
    console.log("4. Agente usando herramientas MCP con Claude...");
    const client = new Anthropic();
    const anthropicTools = tools.map(mcpToolToAnthropicTool);

    const messages: Anthropic.MessageParam[] = [
      {
        role: "user",
        content: "Suma 42 + 58 y luego repite el texto 'Hola MCP'.",
      },
    ];

    let response = await client.messages.create({
      model: MODEL,
      max_tokens: 1024,
      tools: anthropicTools,
      messages,
    });

    // Loop hasta end_turn
    while (response.stop_reason === "tool_use") {
      const toolUseBlocks = response.content.filter(
        (b): b is Anthropic.ToolUseBlock => b.type === "tool_use"
      );

      const toolResults: Anthropic.ToolResultBlockParam[] = [];

      for (const block of toolUseBlocks) {
        console.log(
          `   → ${block.name}(${JSON.stringify(block.input)})`
        );
        const result = await mcp.callTool(
          block.name,
          block.input as Record<string, unknown>
        ) as { content: Array<{ type: string; text: string }> };

        const text = result.content.map((c) => c.text).join("");
        console.log(`   ← ${text}`);

        toolResults.push({
          type: "tool_result",
          tool_use_id: block.id,
          content: text,
        });
      }

      messages.push({ role: "assistant", content: response.content });
      messages.push({ role: "user", content: toolResults });

      response = await client.messages.create({
        model: MODEL,
        max_tokens: 1024,
        tools: anthropicTools,
        messages,
      });
    }

    const finalText = response.content
      .filter((b): b is Anthropic.TextBlock => b.type === "text")
      .map((b) => b.text)
      .join("");

    console.log(`\nRespuesta final: ${finalText}`);
  } finally {
    mcp.close();
    // Limpiar archivo temporal
    try {
      const fs = await import("fs");
      const os = await import("os");
      const path = await import("path");
      fs.unlinkSync(path.join(os.tmpdir(), "mcp-echo-server.js"));
    } catch {
      // ignorar
    }
  }
}

main().catch(console.error);

File: python/05-herramientas/23-mcp.py
# Cliente MCP mínimo.
#
# Implementa el protocolo MCP (Model Context Protocol) via stdio
# para conectar a un servidor MCP, listar sus herramientas y ejecutar tool calls.
#
# El protocolo es JSON-RPC 2.0. El flujo completo:
#   1. initialize (negociar capabilities)
#   2. tools/list (descubrir herramientas disponibles)
#   3. tools/call (ejecutar una herramienta)
#
# Este ejemplo usa un servidor MCP in-process simulado via stdio
# para demostrar el protocolo sin dependencias externas.
#
# Cómo ejecutar:
#   make py SCRIPT=python/05-herramientas/23-mcp.py
#
# Qué esperar:
#   Conexión a un servidor MCP via stdio, listado de herramientas disponibles,
#   y ejecución de un tool call a través del protocolo JSON-RPC 2.0.
#   Si no hay servidor MCP, muestra el protocolo de forma simulada.
#
# Variables de entorno:
#   MODEL — modelo a usar (default: claude-sonnet-4-6)

import json
import asyncio
import tempfile
import os
import anthropic

MODEL = os.environ.get("MODEL", "claude-sonnet-4-6")

# Script del servidor MCP simulado (Python, lanzado como subprocess)
SERVER_SCRIPT = '''
import sys
import json

TOOLS = [
    {
        "name": "echo",
        "description": "Devuelve el texto recibido tal cual.",
        "inputSchema": {
            "type": "object",
            "properties": {"text": {"type": "string"}},
            "required": ["text"]
        }
    },
    {
        "name": "add",
        "description": "Suma dos números.",
        "inputSchema": {
            "type": "object",
            "properties": {
                "a": {"type": "number"},
                "b": {"type": "number"}
            },
            "required": ["a", "b"]
        }
    }
]

for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    try:
        req = json.loads(line)
    except Exception:
        continue

    method = req.get("method", "")
    req_id = req.get("id")

    if method == "initialize":
        resp = {
            "jsonrpc": "2.0", "id": req_id,
            "result": {
                "protocolVersion": "2025-03-26",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "echo-server", "version": "1.0"}
            }
        }
        sys.stdout.write(json.dumps(resp) + "\\n")
        sys.stdout.flush()
        continue

    if method == "notifications/initialized":
        continue

    if method == "tools/list":
        sys.stdout.write(json.dumps({
            "jsonrpc": "2.0", "id": req_id, "result": {"tools": TOOLS}
        }) + "\\n")
        sys.stdout.flush()
        continue

    if method == "tools/call":
        params = req.get("params", {})
        name = params.get("name")
        args = params.get("arguments", {})
        if name == "echo":
            content = [{"type": "text", "text": args.get("text", "")}]
        elif name == "add":
            content = [{"type": "text", "text": str(args.get("a", 0) + args.get("b", 0))}]
        else:
            sys.stdout.write(json.dumps({
                "jsonrpc": "2.0", "id": req_id,
                "error": {"code": -32601, "message": f"Tool not found: {name}"}
            }) + "\\n")
            sys.stdout.flush()
            continue
        sys.stdout.write(json.dumps({
            "jsonrpc": "2.0", "id": req_id,
            "result": {"content": content, "isError": False}
        }) + "\\n")
        sys.stdout.flush()
        continue

    sys.stdout.write(json.dumps({
        "jsonrpc": "2.0", "id": req_id,
        "error": {"code": -32601, "message": f"Method not found: {method}"}
    }) + "\\n")
    sys.stdout.flush()
'''


# --- Cliente MCP mínimo via stdio ---

class McpStdioClient:
    def __init__(self, command: str, args: list[str]):
        self._command = command
        self._args = args
        self._proc = None
        self._next_id = 1
        self._pending: dict[int, asyncio.Future] = {}
        self._reader_task = None

    async def start(self):
        self._proc = await asyncio.create_subprocess_exec(
            self._command, *self._args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=None,
        )
        self._reader_task = asyncio.create_task(self._read_loop())

    async def _read_loop(self):
        while True:
            line = await self._proc.stdout.readline()
            if not line:
                break
            try:
                msg = json.loads(line.decode())
                req_id = msg.get("id")
                fut = self._pending.pop(req_id, None)
                if fut and not fut.done():
                    fut.set_result(msg)
            except Exception:
                pass  # ignorar líneas que no son JSON

    async def _send(self, method: str, params: dict | None = None) -> object:
        req_id = self._next_id
        self._next_id += 1
        req = {"jsonrpc": "2.0", "id": req_id, "method": method}
        if params is not None:
            req["params"] = params

        loop = asyncio.get_event_loop()
        fut = loop.create_future()
        self._pending[req_id] = fut

        self._proc.stdin.write((json.dumps(req) + "\n").encode())
        await self._proc.stdin.drain()

        resp = await fut
        if "error" in resp:
            raise RuntimeError(resp["error"]["message"])
        return resp.get("result")

    async def initialize(self):
        await self._send("initialize", {
            "protocolVersion": "2025-03-26",
            "capabilities": {"tools": {}},
            "clientInfo": {"name": "agente-libro", "version": "1.0"},
        })
        # Notificación (no espera respuesta)
        notif = json.dumps({"jsonrpc": "2.0", "method": "notifications/initialized"}) + "\n"
        self._proc.stdin.write(notif.encode())
        await self._proc.stdin.drain()

    async def list_tools(self) -> list[dict]:
        result = await self._send("tools/list", {})
        return result.get("tools", [])

    async def call_tool(self, name: str, arguments: dict) -> object:
        return await self._send("tools/call", {"name": name, "arguments": arguments})

    async def close(self):
        if self._reader_task:
            self._reader_task.cancel()
        if self._proc:
            self._proc.kill()
            await self._proc.wait()


# --- Convertir herramientas MCP al formato de Anthropic ---

def mcp_tool_to_anthropic(tool: dict) -> dict:
    return {
        "name": tool["name"],
        "description": tool["description"],
        "input_schema": tool["inputSchema"],
    }


# --- Demo completa ---

async def main():
    print("=== Cliente MCP mínimo ===\n")

    # Escribir el script del servidor en un archivo temporal
    with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
        f.write(SERVER_SCRIPT)
        tmp_path = f.name

    mcp = McpStdioClient("python3", [tmp_path])

    try:
        await mcp.start()

        # 1. Inicializar conexión
        print("1. Inicializando conexión MCP...")
        await mcp.initialize()
        print("   OK — handshake completado\n")

        # 2. Listar herramientas disponibles
        print("2. Descubriendo herramientas (tools/list)...")
        tools = await mcp.list_tools()
        print(f"   {len(tools)} herramientas disponibles:")
        for tool in tools:
            print(f"   - {tool['name']}: {tool['description']}")
        print()

        # 3. Ejecutar una tool call directa via MCP
        print("3. Llamada directa: add(17, 25)")
        add_result = await mcp.call_tool("add", {"a": 17, "b": 25})
        text = add_result["content"][0]["text"]
        print(f"   Resultado: {text}\n")

        # 4. Usar las herramientas MCP con el modelo
        print("4. Agente usando herramientas MCP con Claude...")
        client = anthropic.Anthropic()
        anthropic_tools = [mcp_tool_to_anthropic(t) for t in tools]

        messages = [
            {
                "role": "user",
                "content": "Suma 42 + 58 y luego repite el texto 'Hola MCP'.",
            }
        ]

        response = client.messages.create(
            model=MODEL,
            max_tokens=1024,
            tools=anthropic_tools,
            messages=messages,
        )

        # Loop hasta end_turn
        while response.stop_reason == "tool_use":
            tool_blocks = [b for b in response.content if b.type == "tool_use"]
            tool_results = []

            for block in tool_blocks:
                print(f"   → {block.name}({json.dumps(block.input)})")
                result = await mcp.call_tool(block.name, block.input)
                text = "".join(c["text"] for c in result["content"])
                print(f"   ← {text}")
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": text,
                })

            messages.append({"role": "assistant", "content": response.content})
            messages.append({"role": "user", "content": tool_results})

            response = client.messages.create(
                model=MODEL,
                max_tokens=1024,
                tools=anthropic_tools,
                messages=messages,
            )

        final_text = "".join(b.text for b in response.content if b.type == "text")
        print(f"\nRespuesta final: {final_text}")

    finally:
        await mcp.close()
        os.unlink(tmp_path)


if __name__ == "__main__":
    asyncio.run(main())
